File: app/src/main/java/com/ocram/qichwadic/framework/dao/DictionaryDao.java
package com.ocram.qichwadic.framework.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.Dictionary;

import java.util.List;

import javax.inject.Singleton;

import io.reactivex.Flowable;

@Dao
@Singleton
public interface DictionaryDao {

    @Query("SELECT COUNT(*) from dictionary")
    Flowable<Integer> getSavedDictionariesTotal();

    @Query("SELECT d.* from dictionary d order by d.is_quechua DESC, d.id")
    Flowable<List<Dictionary>> getDictionaries();

    @Query("SELECT * from dictionary where is_quechua = :fromQuechua")
    Flowable<List<Dictionary>> getDictionaries(boolean fromQuechua);

    @Query("DELETE from dictionary where id = :id")
    void removeOne(int id);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertDictionaryAndDefinitions(Dictionary dictionary, List<Definition> definitions);

}

File: app/src/main/java/com/ocram/qichwadic/features/common/data/local/dao/DictionaryDao.kt
package com.ocram.qichwadic.features.common.data.local.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query

import com.ocram.qichwadic.features.common.data.model.DefinitionEntity
import com.ocram.qichwadic.features.common.data.model.DictionaryEntity

@Dao
interface DictionaryDao {

    @Query("SELECT d.* from dictionary d order by d.is_quechua DESC, d.id")
    fun getDictionaries(): List<DictionaryEntity>

    @Query("DELETE from dictionary where id = :id")
    fun removeOne(id: Int): Int

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insertDictionaryAndDefinitions(dictionary: DictionaryEntity, definitions: List<DefinitionEntity>)

}
